app/backend/src/couchers/servicers/auth.py

```diff
@@ -424,19 +424,31 @@ def CompletePasswordReset(self, request, context):
             else:
                 context.abort(grpc.StatusCode.NOT_FOUND, errors.INVALID_TOKEN)
 
-    # Only called if user doesn't have password
-    def ConfirmChangeEmailWithOldAddress(self, request, context):
+    def ConfirmChangeEmail(self, request, context):
         with session_scope() as session:
-            user = (
-                session.query(User)
-                .filter(User.old_email_token == request.change_email_token)
+            query = session.query(User).filter(User.id == context.user_id)
+            if (
+                query.filter(User.old_email_token == request.change_email_token)
                 .filter(User.old_email_token_created <= func.now())
                 .filter(User.old_email_token_expiry >= func.now())
-                .one_or_none()
-            )
-            if not user:
+                .one()
+            ):
+                return self.ConfirmChangeEmailWithOldAddress(request, context)
+            elif (
+                query.filter(User.new_email_token == request.change_email_token)
+                .filter(User.new_email_token_created <= func.now())
+                .filter(User.new_email_token_expiry >= func.now())
+                .one()
+            ):
+                return self.ConfirmChangeEmailWithNewAddress(request, context)
+            else:
                 context.abort(grpc.StatusCode.NOT_FOUND, errors.INVALID_TOKEN)
 
+    # Only called if user doesn't have password
+    def ConfirmChangeEmailWithOldAddress(self, request, context):
```

Comment:
You should remove this, right?

Reply:
I need to rejigger a few things. Probably reword, just to emphasize it's not called every time